x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go

```diff
@@ -0,0 +1,264 @@
+func NewAggregator(config AggregatorConfig) (*Aggregator, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid aggregator config")
+	}
+	if config.Logger == nil {
+		config.Logger = logp.NewLogger(logs.ServiceSummaryMetrics)
+	}
+	aggregator := Aggregator{
+		config:   config,
+		active:   make(map[time.Duration]*metricsBuffer),
+		inactive: make(map[time.Duration]*metricsBuffer),
+	}
+	base, err := baseaggregator.New(baseaggregator.AggregatorConfig{
+		PublishFunc:     aggregator.publish, // inject local publish
+		Logger:          config.Logger,
+		Interval:        config.Interval,
+		RollUpIntervals: config.RollUpIntervals,
+	})
+	if err != nil {
+		return nil, err
+	}
+	aggregator.Aggregator = base
+	for _, interval := range aggregator.Intervals {
+		aggregator.active[interval] = newMetricsBuffer(config.MaxGroups)
+		aggregator.inactive[interval] = newMetricsBuffer(config.MaxGroups)
+	}
+	return &aggregator, nil
+}
+
+func (a *Aggregator) publish(ctx context.Context, period time.Duration) error {
+	// We hold a.mu only long enough to swap the serviceSummaryMetrics. This will
+	// be blocked by serviceSummaryMetrics updates, which is OK, as we prefer not
+	// to block serviceSummaryMetrics updaters. After the lock is released nothing
+	// will be accessing a.inactive.
+	a.mu.Lock()
+
+	// EXPLAIN: We swap active <-> inactive, so that we're only working on the
+	// inactive property while publish is running. `a.active` is the buffer that
+	// receives/stores/updates the metricsets, once swapped, we're working on the
+	// `a.inactive` which we're going to process and publish.
+	current := a.active[period]
+	a.active[period], a.inactive[period] = a.inactive[period], current
+	a.mu.Unlock()
+
+	size := len(current.m)
+	if size == 0 {
+		a.config.Logger.Debugf("no service summary metrics to publish")
+		return nil
+	}
+
+	intervalStr := interval.FormatDuration(period)
+	batch := make(model.Batch, 0, size)
+	for key, aggKey := range current.m {
+		m := makeMetricset(aggKey, intervalStr)
+		batch = append(batch, m)
+		delete(current.m, key)
+	}
+	a.config.Logger.Debugf("publishing %d metricsets", len(batch))
+	return a.config.BatchProcessor.ProcessBatch(ctx, &batch)
+}
+
+// ProcessBatch aggregates all service summary metrics.
+func (a *Aggregator) ProcessBatch(ctx context.Context, b *model.Batch) error {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	for _, event := range *b {
+		// Not filtering on Processor.
+		a.processEvent(&event, b)
+	}
+	return nil
+}
+
+func (a *Aggregator) processEvent(event *model.APMEvent, b *model.Batch) {
+	for _, period := range a.Intervals {
+		key := makeAggregationKey(event, period)
+		if !a.active[period].storeOrUpdate(key, a.config.Logger) {
+			intervalStr := interval.FormatDuration(period)
+			*b = append(*b, makeMetricset(key, intervalStr))
+		}
+	}
+}
+
+type metricsBuffer struct {
+	mu sync.RWMutex
+	m  map[uint64]aggregationKey
```

Comment:
pending discussion on hash collision and whether to use a chained hash table like service metrics

Reply:
We definitely need to handle collisions. How that's implemented is less important, but we should probably be consistent and use separate chaining.